egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_decode.py

```diff
@@ -6,83 +6,286 @@
 import torch
 
-from snowfall.data import LibriSpeechAsrDataModule
+from k2 import Fsa, SymbolTable
+
 from snowfall.common import average_checkpoint, store_transcripts
+from snowfall.common import find_first_disambig_symbol
 from snowfall.common import get_texts
 from snowfall.common import load_checkpoint
 from snowfall.common import str2bool
 from snowfall.common import write_error_stats
+from snowfall.data import LibriSpeechAsrDataModule
+from snowfall.decoding.graph import compile_HLG
+from snowfall.decoding.lm_rescore import rescore_with_n_best_list
+from snowfall.decoding.lm_rescore import rescore_with_whole_lattice
+from snowfall.decoding.lm_rescore import compute_am_scores, _intersect_device
+from snowfall.models import AcousticModel
 from snowfall.models.conformer import Conformer
 from snowfall.text.numericalizer import Numericalizer
 from snowfall.training.ctc_graph import build_ctc_topo
+from snowfall.training.mmi_graph import get_phone_symbols
+
+
+def nbest_decoding(model, encoder_memory, memory_mask, lats: k2.Fsa, num_paths: int):
+    '''
+    N-best rescore with transformer-decoder model.
+    The basic idea is to first extra n-best paths from the given lattice.
+    Then extract word_seqs and token_seqs for each path.
+    Compute the negative log-likehood for each token_seq as 'language model score', called decoder_scores.
+    Compute am score for each token_seq.
+    Total scores is a weight sum of am_score and decoder_scores.
+    The one with the max total score is used as the decoding output.
+    '''
+
+    # lats has token IDs as labels
+    # and word IDs as aux_labels.
+    # First, extract `num_paths` paths for each sequence.
+    # paths is a k2.RaggedInt with axes [seq][path][arc_pos]
+    paths = k2.random_paths(lats, num_paths=num_paths, use_double_scores=True)
+
+    # token_seqs/word_seqs is a k2.RaggedInt sharing the same shape as `paths`
+    # but it contains word IDs. Note that it also contains 0s and -1s.
+    # The last entry in each sublist is -1.
+    token_seqs = k2.index(lats.labels.contiguous(), paths)
+    word_seqs = k2.index(lats.aux_labels.contiguous(), paths)
+
+    # Note: the above operation supports also the case when
+    # lats.aux_labels is a ragged tensor. In that case,
+    # `remove_axis=True` is used inside the pybind11 binding code,
+    # so the resulting `word_seqs` still has 3 axes, like `paths`.
+    # The 3 axes are [seq][path][word]
+
+    # Remove epsilons and -1 from word_seqs
+    token_seqs = k2.ragged.remove_values_leq(token_seqs, 0)
+    word_seqs = k2.ragged.remove_values_leq(word_seqs, 0)
+
+    # Remove repeated sequences to avoid redundant computation later.
+    #
+    # Since k2.ragged.unique_sequences will reorder paths within a seq,
+    # `new2old` is a 1-D torch.Tensor mapping from the output path index
+    # to the input path index.
+    # new2old.numel() == unique_word_seqs.num_elements()
+    unique_token_seqs, _, new2old = k2.ragged.unique_sequences(
```

Comment:
Could you use the approach we are using in the current master?

That is, use `unique_word_seqs`, not `unique_token_seqs`, to compute the `lm_scores`.

Different token seqs in `unique_tokens_seqs` may correspond to the same word seqs.
`lm_scores` is for word seqs, not token seqs.

Reply:
will try it. 
Now I use unique_token_seqs rather than unique_word_seqs for because of following two reasons:
1. token_seq is always a 1-to-1 map to word_seq. These should not be many disambiguations.
2. transformer decoder is trained by token_seq. unique_token_seqs is already generated for transformer decoder, so I use it to get lm_scores. 
 
Actually when you want get word_seq from token_seq, just do:
```
word_seq = ''.join(token_seq).replace('_',' ')
```

Reply:
> token_seq is always a 1-to-1 map to word_seq. These should not be many disambiguations.

Are there epsilons (0s) in `token seqs`? Are there contiguous repeated tokens in `token seqs`? 
token seqs from the above two cases can correspond to the same word seq, I think.

----

> transformer decoder is trained by token_seq. unique_token_seqs is already generated for transformer decoder, so I use it to get lm_scores.

Is it possible to get the token seq from a word seq given the word piece model?

Reply:
Fangjun is right that we should the unique_word_seqs, because even though it's a 1-1 map, that won't be obvious to `k2.ragged.unique_sequences`; many of them will really be repeats.  When composing the LM with the CTC topo, we need to keep the "inner_labels" as an attribute, I believe `compose()` has an arg "inner_labels_name" or something like that that so the inner (matched) labels can be kept.